databend-jdbc/src/main/java/com/databend/jdbc/DatabendConnection.java

```diff
@@ -702,17 +690,42 @@ DatabendClient startQueryWithFailover(String sql, StageAttachment attach) throws
                 ClientSettings s = sb.build();
                 logger.log(Level.FINE, "retry " + i + " times to execute query: " + sql + " on " + s.getHost());
                 // discover new hosts in need.
-//                if (this.autoDiscovery) {
-//
-//                }
+                if (this.autoDiscovery) {
+                    tryAutoDiscovery(httpClient, s);
+                }
                 return new DatabendClientV1(httpClient, sql, s, this);
             } catch (RuntimeException e1) {
                 e = e1;
             } catch (Exception e1) {
                 throw new SQLException("Error executing query: " + "SQL: " + sql + " " + e1.getMessage() + " cause: " + e1.getCause(), e1);
             }
         }
-        throw new SQLException("Failover Retry Error executing query after" + getMaxFailoverRetries() +  "failover retry: " + "SQL: " + sql + " " + e.getMessage() + " cause: " + e.getCause(), e);
+        throw new SQLException("Failover Retry Error executing query after " + getMaxFailoverRetries() + " failover retry: " + "SQL: " + sql + " " + e.getMessage() + " cause: " + e.getCause(), e);
+    }
+
+    /**
+     * Try to auto discovery the databend nodes it will log exceptions when auto discovery failed and not affect real query execution
+     *
+     * @param client the http client to query on
+     * @param settings the client settings to use
+     */
+    void tryAutoDiscovery(OkHttpClient client, ClientSettings settings) {
+        if (this.autoDiscovery) {
+            if (this.driverUri.enableMock()) {
+                settings.getAdditionalHeaders().put("~mock.unsupported.discovery", "true");
+            }
+            DatabendNodes nodes = this.driverUri.getNodes();
+            if (nodes != null && nodes.needDiscovery()) {
+                try {
+                    nodes.discoverUris(client, settings);
+                } catch (UnsupportedOperationException e) {
+                    logger.log(Level.WARNING, "Current Query Node do not support auto discovery, close the functionality: " + e.getMessage());
+                    this.autoDiscovery = false;
+                } catch (Exception e) {
+                    logger.log(Level.FINE, "Error auto discovery: " + " cause: " + e.getCause() + " message: " + e.getMessage());
```

Comment:
first time to find out FINE log level, what's it's normal use case? 

(not an issue at all, just curious 👀)

Reply:
info level will print problem by default, I think it is similar to DEBUG level on golang